azure/services/virtualmachines/virtualmachines.go

```diff
@@ -138,6 +153,35 @@ func (s *Service) Delete(ctx context.Context) error {
 	return err
 }
 
+func (s *Service) checkUserAssignedIdentities(ctx context.Context, specIdentities []infrav1.UserAssignedIdentity, vmIdentities []infrav1.UserAssignedIdentity) error {
+	expectedMap := make(map[string]int)
+	actualMap := make(map[string]int)
+
+	// Create a map of the expected identities. The ProviderID is converted to match the format of the VM identity.
+	for _, expectedIdentity := range specIdentities {
+		expectedClientID, err := s.identitiesGetter.GetClientID(ctx, expectedIdentity.ProviderID)
+		if err != nil {
+			return errors.Wrap(err, "failed to get client ID")
+		}
+		expectedMap[expectedClientID]++
+	}
+
+	// Create a map of the actual identities from the vm.
+	for _, actualIdentity := range vmIdentities {
+		actualMap[actualIdentity.ProviderID]++
+	}
+
+	// Check if the expected identities are present in the vm.
+	for expectedKey, expectedVal := range expectedMap {
+		if actualMap[expectedKey] != expectedVal {
```

Comment:
I'm not familiar with the semantics around specifying the same identity more than once (or if that's even possible), so this is kind of a spitball, but would it be acceptable to have the same identity listed more times in the actual VM than in the CAPZ spec?
```suggestion
		if actualMap[expectedKey] < expectedVal {
```

Reply:
I don't think it's allowed to have the same identity more than once, but I implemented this using maps for efficiency. However, it would make more sense to use `<` since we are always seeing if the identity is missing.

Reply:
If that's the case, I think using `map[string]struct{}` might make the semantics a bit more clear.

Reply:
That makes sense! What would I add to the struct to indicate that the identity exists? Would I just add the identity itself?

Reply:
`struct{}` by itself is a complete type where `struct{}{}` is the only possible value. Then the identity is used as the key and can be checked with something like:
```go
_, exists := actualMap[expectedKey]
```